From 99a94660f730397e45fc933592aa5f6aa4cea9dc Mon Sep 17 00:00:00 2001
From: Jan Grulich <jgrulich@redhat.com>
Date: Wed, 25 Mar 2020 13:09:37 +0100
Subject: OpenURI: open file with write access when requested

When user specifies to open a file with write access, we need to open it with specified
permission, otherwise xdg-desktop-portal will fail as it checks for write access.

Also follow glib and do not use O_PATH, because since xdg-desktop-portal 1.0.1, regular file descriptors
are also accepted.

diff --git a/libportal/openuri.c b/libportal/openuri.c
index dfcd0f4..9f3535e 100644
--- a/libportal/openuri.c
+++ b/libportal/openuri.c
@@ -209,11 +209,16 @@ do_open (OpenCall *call)
     {
       g_autoptr(GUnixFDList) fd_list = NULL;
       g_autofree char *path = NULL;
-      int fd, fd_in;
+      int fd, fd_in, flags;
 
       path = g_file_get_path (file);
 
-      fd = g_open (path, O_PATH | O_CLOEXEC);
+      if (call->writable)
+        flags = O_RDWR | O_CLOEXEC;
+      else
+        flags = O_RDONLY | O_CLOEXEC;
+
+      fd = g_open (path, flags);
       if (fd == -1)
         {
           g_task_return_new_error (call->task, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to open '%s'", call->uri);
